README
NAME
    "Net::Async::WebSocket" - use WebSockets with "IO::Async"

DESCRIPTION
    This distribution provides modules that implement the WebSocket
    protocol, and allows either servers or clients to be written based on
    IO::Async.

    To implement a server, see Net::Async::WebSocket::Server.

    To implement a client, see Net::Async::WebSocket::Client.

    This module itself does not provide any code, and exists only to provide
    the module $VERSION and top-level documentation.

SEE ALSO
    *       Protocol::WebSocket - WebSocket protocol

AUTHOR
    Paul Evans <leonerd@leonerd.org.uk>



Changes
Revision history for Net-Async-WebSocket

0.08    2014/03/27 16:43:22
        [CHANGES]
         * First-class support for Futures
         * Subclass IO::Async::Stream directly; avoid IaProtocol::Stream

        [BUGFIXES]
         * Only pull out ->configure arguments that actually exist
         * Handle defined-but-false frames (RT92477)

0.07    CHANGES:
         * ->socketpair now comes from IO::Async::OS, not the underlying loop

0.06    CHANGES:
         * Added on_handshake test/continuation for early-aborting connections
           if required (patch by David Leadbeater) (RT#83436)

0.05    CHANGES:
         * Pass all arguments from send_frame into Protocol::WebSocket::Frame
           constructor (RT79202)

0.04    CHANGES:
         * Updated for Protocol::WebSocket 0.009 - s/to_string/to_bytes/

0.03    CHANGES:
         * Use IO::Async::Protocol->connect from 0.34

0.02    CHANGES:
         * Bugfixes for getaddrinfo that isn't on Linux
         * Created some example scripts and SYNOPSIS code samples

0.01    First version, released on an unsuspecting world.



lib/Net/Async/WebSocket.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2010 -- leonerd@leonerd.org.uk

package Net::Async::WebSocket;

use strict;
use warnings;

our $VERSION = '0.08';

=head1 NAME

C<Net::Async::WebSocket> - use WebSockets with C<IO::Async>

=head1 DESCRIPTION

This distribution provides modules that implement the WebSocket protocol, and
allows either servers or clients to be written based on L<IO::Async>.

To implement a server, see L<Net::Async::WebSocket::Server>.

To implement a client, see L<Net::Async::WebSocket::Client>.

This module itself does not provide any code, and exists only to provide the
module C<$VERSION> and top-level documentation.

=cut

=head1 SEE ALSO

=over 8

=item *

L<Protocol::WebSocket> - WebSocket protocol

=back

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/Net/Async/WebSocket/Client.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2010-2014 -- leonerd@leonerd.org.uk

package Net::Async::WebSocket::Client;

use strict;
use warnings;
use base qw( Net::Async::WebSocket::Protocol );

use Carp;

our $VERSION = '0.08';

use Protocol::WebSocket::Handshake::Client;

=head1 NAME

C<Net::Async::WebSocket::Client> - connect to a WebSocket server using
C<IO::Async>

=head1 SYNOPSIS

 use IO::Async::Loop;
 use Net::Async::WebSocket::Client;

 my $client = Net::Async::WebSocket::Client->new(
    on_frame => sub {
       my ( $self, $frame ) = @_;
       print $frame;
    },
 );

 my $loop = IO::Async::Loop->new;
 $loop->add( $client );
 
 $client->connect(
    host => $HOST,
    service => $PORT,
    url => "ws://$HOST:$PORT/",

    on_connected => sub {
       $client->send_frame( "Hello, world!\n" );
    },

    on_connect_error => sub { die "Cannot connect - $_[-1]" },
    on_resolve_error => sub { die "Cannot resolve - $_[-1]" },
 );

 $loop->loop_forever;

=head1 DESCRIPTION

This subclass of L<Net::Async::WebSocket::Protocol> connects to a WebSocket
server to establish a WebSocket connection for passing frames.

=cut

=head1 METHODS

=cut

sub _do_handshake
{
   my $self = shift;
   my %params = @_;

   my $hs = Protocol::WebSocket::Handshake::Client->new(
      url => $params{url},
   );

   $self->debug_printf( "HANDSHAKE start" );
   $self->write( $hs->to_string );

   my $f = $self->loop->new_future;
   $self->SUPER::configure( on_read => sub {
      my ( undef, $buffref, $closed ) = @_;

      $hs->parse( $$buffref ); # modifies $$buffref

      if( $hs->is_done ) {
         $self->debug_printf( "HANDSHAKE done" );
         $self->SUPER::configure( on_read => undef );

         $f->done( $self );
      }

      return 0;
   } );

   return $f;
}

=head2 $self->connect( %params ) ==> ( $self )

Connect to a WebSocket server. Takes the following named parameters:

=over 8

=item url => STRING

URL to provide to WebSocket handshake

=back

=head2 $self->connect( %params )

When not returning a C<Future>, the following additional parameters provide
continuations:

=over 8

=item on_connected => CODE

CODE reference to invoke when the handshaking is complete.

=back

=cut

sub connect
{
   my $self = shift;
   my %params = @_;

   my $f = $self->SUPER::connect( %params )->then( sub {
      my ( $self ) = @_;

      $self->_do_handshake( %params );
   });

   $f->on_done( $params{on_connected} ) if $params{on_connected};

   return $f if defined wantarray;

   $f->on_ready( sub { undef $f } ); # intentional cycle
}

=head2 $client->connect_handle( $handle, %params ) ==> ( $self )

Sets the read and write handles to the IO reference given, then performs the
initial handshake using the parameters given. These are as for C<connect>.

=cut

sub connect_handle
{
   my $self = shift;
   my ( $handle, %params ) = @_;

   $self->set_handle( $handle );

   $self->_do_handshake( %params );
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/Net/Async/WebSocket/Protocol.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2010-2014 -- leonerd@leonerd.org.uk

package Net::Async::WebSocket::Protocol;

use strict;
use warnings;
use base qw( IO::Async::Stream );

use Carp;

our $VERSION = '0.08';

use Protocol::WebSocket::Frame;

=head1 NAME

C<Net::Async::WebSocket::Protocol> - send and receive WebSocket frames

=head1 DESCRIPTION

This subclass of L<IO::Async::Stream> implements an established WebSocket
connection, that has already completed its setup handshaking and is ready to
pass frames.

Objects of this type would not normally be constructed directly. For WebSocket
clients, see L<Net::Async::WebSocket::Client>, which is a subclass of this.
For WebSocket servers, see L<Net::Async::WebSocket::Server>, which constructs
objects in this class when it accepts a new connection and passes it to its
event handler.

=cut

sub _init
{
   my $self = shift;
   $self->SUPER::_init;

   $self->{framebuffer} = Protocol::WebSocket::Frame->new;
}

=head1 PARAMETERS

The following named parameters may be passed to C<new> or C<configure>:

=over 8

=item on_frame => CODE

A CODE reference for when a frame is received

 $on_frame->( $self, $frame )

=back

=cut

sub configure
{
   my $self = shift;
   my %params = @_;

   foreach (qw( on_frame )) {
      $self->{$_} = delete $params{$_} if exists $params{on_frame};
   }

   $self->SUPER::configure( %params );
}

my %FRAMETYPES = (
   1 => "text",
   2 => "binary",
   0x8 => "close",
   0x9 => "ping",
   0xa => "pong",
);

sub on_read
{
   my $self = shift;
   my ( $buffref, $closed ) = @_;

   my $framebuffer = $self->{framebuffer};

   $framebuffer->append( $$buffref ); # modifies $$buffref

   while( defined( my $frame = $framebuffer->next ) ) {
      $self->debug_printf( "FRAME " . $FRAMETYPES{$framebuffer->opcode} );

      $self->invoke_event( on_frame => $frame );
   }

   return 0;
}

=head1 METHODS

=cut

=head2 $self->send_frame( @args )

Sends a frame to the peer containing containing the given string. The
arguments are passed to L<Protocol::WebSocket::Frame>'s C<new> method.

=cut

sub send_frame
{
   my $self = shift;

   $self->write( Protocol::WebSocket::Frame->new( @_ )->to_bytes );
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;


lib/Net/Async/WebSocket/Server.pm
#  You may distribute under the terms of either the GNU General Public License
#  or the Artistic License (the same terms as Perl itself)
#
#  (C) Paul Evans, 2010-2014 -- leonerd@leonerd.org.uk

package Net::Async::WebSocket::Server;

use strict;
use warnings;
use base qw( IO::Async::Listener );

use Carp;

our $VERSION = '0.08';

use Net::Async::WebSocket::Protocol;

use Protocol::WebSocket::Handshake::Server;

=head1 NAME

C<Net::Async::WebSocket::Server> - serve WebSocket clients using C<IO::Async>

=head1 SYNOPSIS

 use IO::Async::Loop;
 use Net::Async::WebSocket::Server;
 
 my $server = Net::Async::WebSocket::Server->new(
    on_client => sub {
       my ( undef, $client ) = @_;
 
       $client->configure(
          on_frame => sub {
             my ( $self, $frame ) = @_;
             $self->send_frame( $frame );
          },
       );
    }
 );
 
 my $loop = IO::Async::Loop->new;
 $loop->add( $server );
 
 $server->listen(
    service => 3000,
 
    on_listen_error => sub { die "Cannot listen - $_[-1]" },
    on_resolve_error => sub { die "Cannot resolve - $_[-1]" },
 );
 
 $loop->loop_forever;

=head1 DESCRIPTION

This subclass of L<IO::Async::Listener> accepts WebSocket connections. When a
new connection arrives it will perform an initial handshake, and then pass the
connection on to the continuation callback or method.

=cut

sub new
{
   my $class = shift;
   return $class->SUPER::new(
      @_,
      handle_class => "Net::Async::WebSocket::Protocol",
      on_accept => sub {
         my ( $self, $client ) = @_;

         my $hs = Protocol::WebSocket::Handshake::Server->new;

         $client->configure(
            on_read => sub {
               my ( $client, $buffref, $closed ) = @_;

               $hs->parse( $$buffref ); # modifies $$buffref

               if( $hs->is_done ) {
                  my $on_handshake = $self->can_event( "on_handshake" ) ||
                     sub { $_[3]->( 1 ) };

                  $on_handshake->( $self, $client, $hs, sub {
                     my ( $ok ) = @_;

                     unless( $ok ) {
                        $self->remove_child( $client );
                        return;
                     }

                     $client->configure( on_read => undef );
                     $client->write( $hs->to_string );

                     $client->debug_printf( "HANDSHAKE done" );
                     $self->invoke_event( on_client => $client );
                  } );
               }

               return 0;
            },
         );

         $self->add_child( $client );
      },
   );
}

=head1 PARAMETERS

The following named parameters may be passed to C<new> or C<configure>:

=over 8

=item on_client => CODE

A callback that is invoked whenever a new client connects and completes its
inital handshake.

 $on_client->( $self, $client )

It will be passed a new instance of a L<Net::Async::WebSocket::Protocol>
object, wrapping the client connection.

=item on_handshake => CODE

A callback that is invoked when a handshake has been requested.

 $on_handshake->( $self, $client, $hs, $continuation )

Calling C<$continuation> with a true value will complete the handshake, false
will drop the connection.

This is useful for filtering on origin, for example:

 on_handshake => sub {
    my ( $self, $client, $hs, $continuation ) = @_;

    $continuation->( $hs->req->origin eq "http://localhost" );
 }

=back

=cut

sub configure
{
   my $self = shift;
   my %params = @_;

   foreach (qw( on_client on_handshake )) {
      $self->{$_} = delete $params{$_} if exists $params{$_};
   }

   $self->SUPER::configure( %params );
}

sub listen
{
   my $self = shift;
   my %params = @_;

   $self->SUPER::listen(
      socktype => 'stream',
      %params,
   );
}

=head1 AUTHOR

Paul Evans <leonerd@leonerd.org.uk>

=cut

0x55AA;
